File: frontend/src/types/pipeline.ts
export type FieldType = 'text' | 'integer' | 'decimal' | 'boolean' | 'select' | 'file' | 'folder';
export type StepStatus = 'pending' | 'scheduled' | 'running' | 'ok' | 'error' | 'cancelled';
export type IssueSeverity = 'warning' | 'blocker';

export interface ValueSpec {
  key: string;
  label?: string;
  type: FieldType;
  required?: boolean;
  default?: unknown;
  options?: string[];
  min?: number;
  max?: number;
  description?: string;
  source_step?: string;
  source_output?: string;
}

export interface OutputSpec {
  key?: string;
  label?: string;
  path: string;
}

export interface PipelineStep {
  id: string;
  name: string;
  description?: string;
  command?: string;
  environment?: string;
  working_directory?: string;
  inputs: ValueSpec[];
  outputs: OutputSpec[];
  parameters: ValueSpec[];
  dependencies: string[];
}

export interface PipelineConfig {
  steps: PipelineStep[];
}

export interface ValidationIssue {
  severity: IssueSeverity;
  message: string;
  step_id?: string;
  field?: string;
}

export interface StepRuntimeState {
  status: StepStatus;
  selected: boolean;
  exit_code?: number | null;
  last_run_at?: string | null;
  message?: string;
}

export interface ExecutionStatus {
  running: boolean;
  current_step_id?: string | null;
  stop_requested: boolean;
}

export interface ProjectSnapshot {
  project_path: string;
  pipeline?: PipelineConfig | null;
  params: Record<string, Record<string, unknown>>;
  state: Record<string, StepRuntimeState>;
  visual_layout: Record<string, { x: number; y: number }>;
  validation: ValidationIssue[];
}

export interface TreeNode {
  name: string;
  path: string;
  type: 'file' | 'directory';
  children: TreeNode[];
}

export interface FilePreview {
  path: string;
  type: 'text' | 'image' | 'binary' | 'missing';
  content?: string | null;
  media_type?: string | null;
}

export interface DirectoryEntry {
  name: string;
  path: string;
}

export interface DirectoryListing {
  path: string;
  parent?: string | null;
  directories: DirectoryEntry[];
}

export interface FileEntry {
  name: string;
  path: string;
  type: 'file';
}

export interface FileListing {
  path: string;
  parent?: string | null;
  directories: DirectoryEntry[];
  files: FileEntry[];
}

export interface StepDetail {
  step: PipelineStep;
  params: Record<string, unknown>;
  state: StepRuntimeState;
  validation: ValidationIssue[];
}

export interface StepCreatePayload {
  id: string;
  name: string;
  description?: string;
  command?: string;
  environment?: string;
  working_directory?: string;
  inputs?: ValueSpec[];
  outputs?: OutputSpec[];
  parameters?: ValueSpec[];
  dependencies?: string[];
}

File: backend/app/models/pipeline.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


FieldType = Literal["text", "integer", "decimal", "boolean", "select", "file", "folder"]
IssueSeverity = Literal["warning", "blocker"]


class ValueSpec(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    key: str
    label: str | None = None
    type: FieldType = "text"
    required: bool = False
    default: Any = None
    options: list[str] = Field(default_factory=list)
    minimum: float | None = Field(default=None, alias="min")
    maximum: float | None = Field(default=None, alias="max")
    description: str = ""
    source_step: str | None = None
    source_output: str | None = None

    @field_validator("type", mode="before")
    @classmethod
    def normalize_type(cls, value: Any) -> Any:
        if value == "selector":
            return "select"
        return value


class InputSpec(ValueSpec):
    pass


class ParameterSpec(ValueSpec):
    pass


class OutputSpec(BaseModel):
    key: str = ""
    label: str | None = None
    path: str


class PipelineStep(BaseModel):
    id: str
    name: str
    description: str = ""
    command: str = ""
    environment: str = ""
    working_directory: str = "."
    inputs: list[InputSpec] = Field(default_factory=list)
    outputs: list[OutputSpec] = Field(default_factory=list)
    parameters: list[ParameterSpec] = Field(default_factory=list)
    dependencies: list[str] = Field(default_factory=list)

    def default_values(self) -> dict[str, Any]:
        values: dict[str, Any] = {}
        for item in [*self.inputs, *self.parameters]:
            if item.default is not None:
                values[item.key] = item.default
        for output in self.outputs:
            if output.key:
                values[output.key] = output.path
        return values


class PipelineConfig(BaseModel):
    steps: list[PipelineStep] = Field(default_factory=list)

    def step_by_id(self) -> dict[str, PipelineStep]:
        return {step.id: step for step in self.steps}


class ValidationIssue(BaseModel):
    severity: IssueSeverity
    message: str
    step_id: str | None = None
    field: str | None = None
